### PR TITLE
Changes for 0.2.0

### PR DESCRIPTION
**Pull request by levic**
Major changes:
- add documentation
- django 3.2 support
- add `CSV_PERMISSIONS_RESOLVE_EVALUATORS` to allow defining your own CSV rules

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,253 @@
...
+
+* Every user is expected to have a `user_type` attribute (the equivalent of `django.contrib.auth`'s `Group`)
+
+* A CSV file that defines a permission matrix is used to define what permissions each `user_type` has
+
+### Installation
+
+In your django settings:
+
+* Add `csv_permissions` to `INSTALLED_APPS` (TODO: necessary?)
```
Comment (davecoates) [inline README.md]: I don't think it's necessary right now but I think it's fine to have this in the default instructions anyway. If we add things that require it down the track then it's already there.

Right now we have an AppConfig so that would never be loaded if we don't do this but it doesn't look like it does anything anyway

File: csv_permissions/permissions.py [hunk 2]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
...
+    resolve_permission_name: Optional[str],
+    resolve_evaluators: Iterable[Union[str, ResolveEvaluatorFunc]],
+) -> Tuple[
+    Dict[PermName, Dict[UserType, Evaluator]],
+    Dict[PermName, bool],
+    Set[str],
+    Set[str]
+]:
     """
     :param file_path: Path to the CSV from which to import.
```
Comment (davecoates) [inline csv_permissions/permissions.py]: should be updated to `file_paths`

File: README.md [hunk 3]
```diff
@@ -1,4 +1,253 @@
...
+
+* Add `csv_permissions.permissions.CSVPermissionsBackend` to `AUTHENTICATION_BACKENDS` 
+  
+* Add a `CSV_PERMISSIONS_PATHS` which is an array/tuple of `str`/`pathlib.Path`
```
Comment (davecoates) [inline README.md]: oh this is new. does it merge the permissions from each csv? should clarify that somewhere in the docs

File: README.md [hunk 4]
```diff
@@ -1,4 +1,253 @@
...
+print(
+    "foo-bar is a global permission"
+    if CSVPermissionBackend().is_global("foo-bar")
+    else "foo-bar is a per-object permission"
+)
+```
+
+### Custom Evaluators
+
+By default putting other than a built-in evaluator in a CSV permissions file
```
Comment (davecoates) [inline README.md]: putting anything

File: README.md [hunk 5]
```diff
@@ -1,4 +1,253 @@
...
+
+# The model column can be blank:
+
+,          library,       report_outstanding,yes,      yes,   yes,
+,          library,       report_popularity, yes,      yes,   yes,
+```
+
+The first 4 columns define the permission details:
+
+**Model** is used to resolve the permission name but is otherwise not used. There is no checks that objects passed to the `has_perm()` actually match the correct type.
```
Comment (davecoates) [inline README.md:104]: at this point I'm wondering how permissions are resolved - is it worth mentioning something here? what about when this value is not set?

File: README.md [hunk 6]
```diff
@@ -1,4 +1,253 @@
...
+
+**Evaluators**
+
+The next columns define permission "evaluators".
+
+Built-in evaluators are:
+
+* `all` - user has permission for all objects. Will raise an error  if an object is not passed to `has_perm()`
+* `yes` - user has permission globally. Will raise an error if an object is passed to `has_perm()`.
+* (empty cell) -- user does not have permission (global or per-object)
```
Comment (davecoates) [inline README.md]: so the `custom:` prefix is no longer required for custom rules?

File: README.md [hunk 7]
```diff
@@ -1,4 +1,253 @@
...
+    # normally if nothing matches an exception will be thrown however it 
+    # can be more convenient (especially in early phases of development )
+    # to issue a warning during CSV parsing, and then throw a
+    # NotImplementedError() when the permission is evaluated
+    'csv_permissions.evaluators.resolve_fallback_not_implemented_evaluator',
+)
+
+# if you don't have any customisations you can point to a list/tuple
+# that is defined elsewhere; this is a basic set:
+#CSV_PERMISSIONS_RESOLVE_EVALUATORS = "csv_permissions.evaluators.default_resolve_evaluators"
```
Comment (davecoates) [inline README.md:189]: so you need to define this even if you have no customisations? or can you leave out `CSV_PERMISSIONS_RESOLVE_EVALUATORS` and it defaults automatically?

File: README.md (hunk 5 above)
Comment (davecoates) [reply README.md:104]: Just read the `Permission names` below about how permissions are resolved - maybe just link to that from here?

File: README.md [hunk 8]
```diff
@@ -1,4 +1,253 @@
...
+
+The first 4 columns define the permission details:
+
+**Model** is used to resolve the permission name but is otherwise not used. There is no checks that objects passed to the `has_perm()` actually match the correct type.
+
+**App** is used to resolve the permission name and model.
+
+**Action** is an arbitrary identifier that is used to resolve the permission name.
+
+**Is Global** whether the permission is global or per-object (see "Global Permission" section below)
```
Comment (davecoates) [inline README.md]: it looks like we validate that a permission must be global if Model is not defined. 2 things

1) Is this a hard requirement? could you have a case where you want a custom non-model permission that does accept an object of some kind?

2) if it is a hard requirement should mention it here somewhere

File: csv_permissions/evaluators.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        raise NotImplementedError(message)
+    return evaluate_not_implemented
+
+
+def resolve_validation_evaluator(details: UnresolvedEvaluator) -> Optional[Evaluator]:
+    """
+    This doesn't actually resolve to any evaluators, it just performs some validation
+    """
+    if details.model is None and not details.is_global:
+        raise ValueError("Permissions without Models must be global.")
```
Comment (davecoates) [inline csv_permissions/evaluators.py:22]: as noted above - is this a hard requirement? does it matter if you wanted to use a non-model permission as an object level permission with some arbitrary object?

File: csv_permissions/evaluators.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+def resolve_yes_evaluator(details: UnresolvedEvaluator) -> Optional[Evaluator]:
+    if details.evaluator_name == "yes":
+        if not details.is_global:
+            raise ValueError("'yes' cannot be used as an object-level permission.")
+        return evaluate_yes
+    return None
+
+
+def evaluate_no(user, obj=None) -> bool:
```
Comment (davecoates) [inline csv_permissions/evaluators.py:54]: for consistency with `yes` should we also accept `no` in the csv rather than just empty string? maybe it's more readable to be able to scan across a row and see yes,no,no rather than yes,,

File: csv_permissions/legacy.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    "create": True,
+    "add": True,
+    "change": False,
+    "update": False,
+    "delete": False,
+}
+
+
+def resolve_legacy_validate_global_evaluator(details: UnresolvedEvaluator) -> Optional[Evaluator]:
+    global _PREDEFINED_ACTION_IS_GLOBAL
```
Comment (davecoates) [inline csv_permissions/legacy.py]: why is this necessary here?

File: csv_permissions/permissions.py [hunk 12]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
...
+                perm_user_type_unresolved[permission] = {}
+
+            for i, user_type in enumerate(user_type_headers):
+                try:
+                    evaluator_name = row[prelim_header_count + i]
+                except IndexError:
+                    continue
+
+                if user_type == "":
+                    # if a column has an empty user type then that's allowed but only if the entire column is empty
```
Comment (davecoates) [inline csv_permissions/permissions.py:141]: what's the reason for an empty column?

File: csv_permissions/permissions.py [hunk 13]
```diff
@@ -354,15 +321,28 @@ def is_global_perm(self, perm: str) -> bool:
 
     def has_perm(self, user: Model, perm: str, obj: Model) -> bool:
         try:
-            user_group = str(user.get_profile().user_type)
+            # TODO: don't hardcode this
+            user_type = str(user.get_profile().user_type)
```
Comment (davecoates) [inline csv_permissions/permissions.py]: apart from this it seems like csv_permissions is compatible with any user implementation?

what about just a `resolve_user_type` setting?

File: README.md (hunk 1 above)
Comment (fanglong-alliance) [inline README.md]: just 2 confirm: still TODO?

File: README.md [hunk 14]
```diff
@@ -1,4 +1,253 @@
...
+Loan,      library,       add,               yes,       yes,   yes,       yes,
+Loan,      library,       view,              no,        all,   all,       own,
+Loan,      library,       change,            no,        all,   all,
+Loan,      library,       delete,            no,        all,   all,
+
+# The model column can be blank:
+
+,          library,       report_outstanding,yes,      yes,   yes,
+,          library,       report_popularity, yes,      yes,   yes,
+```
```
Comment (fanglong-alliance) [inline README.md:99]: could use a line here to explicitly state that spacing's not necessary

File: README.md [hunk 15]
```diff
@@ -1,4 +1,253 @@
...
+
+**App** is used to resolve the permission name and model.
+
+**Action** is an arbitrary identifier that is used to resolve the permission name.
+
+**Is Global** whether the permission is global or per-object (see "Global Permission" section below)
+
+**Evaluators**
+
+The next columns define permission "evaluators".
```
Comment (fanglong-alliance) [inline README.md]: > evaluators for `user_type`s respectively

? just to be clear what those items are (`user_type`'s only got a "you need this" mention b4 this point)

File: README.md [hunk 16]
```diff
@@ -1,4 +1,253 @@
...
+
+Built-in evaluators are:
+
+* `all` - user has permission for all objects. Will raise an error  if an object is not passed to `has_perm()`
```
Comment (fanglong-alliance) [inline README.md]: extra space between `error` and `if`

File: README.md [hunk 17]
```diff
@@ -1,4 +1,253 @@
...
+* `yes` - user has permission globally. Will raise an error if an object is passed to `has_perm()`.
```
Comment (fanglong-alliance) [inline README.md:121]: should explain how `yes` is different from `all` and ideally give example (at least mark this's linked to global permissions below and defer to that)

edit: actually think it might be easier to understand by breaking this section into "global=yes evaluators" and "global=no evaluators"

File: README.md [hunk 18]
```diff
@@ -1,4 +1,253 @@
...
+* (empty cell) -- user does not have permission (global or per-object) 
+
```
Comment (fanglong-alliance) [inline README.md:127]: just a generic comment. at times i find it myself sometimes wanting all cells to be filled (optionally) ie. instead of making it empty, put a "no" there, as its hugely helpful for the purpose of counting comma when you're doing a cat in CLI


(jk: can be even better if they line up! all we need's evaluators.parliment where instead of yes and empty you say `yay` and `nay`!)

File: README.md [hunk 19]
```diff
@@ -1,4 +1,253 @@
...
+# that is defined elsewhere; this is a basic set:
+#CSV_PERMISSIONS_RESOLVE_EVALUATORS = "csv_permissions.evaluators.default_resolve_evaluators"
+
+# for compatibility with csv_permissions 0.1.0
+#CSV_PERMISSIONS_RESOLVE_EVALUATORS = "csv_permissions.legacy.resolve_evaluators"
+
+```
+
+The following code will define some custom evaluators: 
+- `'if_monday'` grants all access on mondays.
```
Comment (fanglong-alliance) [inline README.md:197]: 🤣 🤣 🤣 🤣 🤣 🤣 🤣 🤣 🤣

File: README.md [hunk 20]
```diff
@@ -1,4 +1,253 @@
...
+        return False
+     
+def resolve_evaluators(details: UnresolvedEvaluator) -> Optional[Evaluator]:
+    if details.evaluator_name == "if_monday":
+        return evaluate_if_monday
+
+    if details.evaluator_name == "all_caps":
+        if details.is_global != False:
+            raise ValueError("'all_caps' cannot be used as a global permission.")
+        return evaluate_if_monday
```
Comment (fanglong-alliance) [inline README.md]: wrong one (yeah i know its "just" readme...)

File: README.md [hunk 21]
```diff
@@ -1,4 +1,253 @@
...
+CSV_PERMISSIONS_RESOLVE_PERM_NAME = 'my_site.auth.resolve_rule_name'
+```
+
+In `my_site/auth.py`:
+```python
+from typing import Optional
+from typing import Type
+
+from django.apps import AppConfig
+from django.apps import apps
```
Comment (fanglong-alliance) [inline README.md]: i think if someone copypaste this they'll get an unused import here

File: README.md [hunk 22]
```diff
@@ -1,4 +1,253 @@
...
+from django.db.models import Model
+
+def resolve_rule_name(app_config: AppConfig, model: Optional[Type[Model]], action: str, is_global: bool) -> str:
+    # here's an implementation that is almost the same as django, but
+    # uses - as a separator instead of _ and .
+    #
+    # we also need to handle with the case where a permission has no associated model
+    if model is None:
+        f"{app_config.label}-{action}"
```
Comment (fanglong-alliance) [inline README.md]: and return

File: README.md (hunk 6 above)
Comment (fanglong-alliance) [inline README.md]: where's `own`?

File: README.md [hunk 23]
```diff
@@ -1,4 +1,253 @@
...
+    else:
+        return f"{app_config.label}-{action}-{model._meta.model_name}"
+
+```
```
Comment (fanglong-alliance) [inline README.md:316]: overall speaking i believe it could benefit from assuming original dev have little knowledge of how _django permission_ works and thus provide one more block explaining the **meaning** of global permissions, when to set it to "yes", and what the impacts are

Comment (fanglong-alliance) [reply README.md:316]: also somewhat related, the model-less permission's shown in the example csv but not explained anywhere above

File: README.md (hunk 6 above)
Comment (fanglong-alliance) [reply README.md]: ah, own's now obsolete and not included by default. shown in example.csv; maybe add a comment line there pointing out that its a nonstandard evaluator

File: csv_permissions/legacy.py [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+
+
+def resolve_legacy_warning_evaluator(details: UnresolvedEvaluator) -> Optional[Evaluator]:
+    # we deliberately make bad permissions raise warnings instead of exceptions so that you can get
+    # up & running quickly with scaffolding an app and then come back and fix the permissions later
+    warnings.warn(
+        "csv_permissions.legacy.legacy_resolve_evaluator() will be removed in future. "
+        "You should copy legacy evaluators into your project.",
+        DeprecationWarning,
+    )
```
Comment (fanglong-alliance) [inline csv_permissions/legacy.py:25]: (skipping this file in CR assuming its just the old one except this warning)

File: csv_permissions/permissions.py [hunk 25]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
         'custom: name_of_custom_rule_function' (has permission as defined by name_of_custom_rule_function)
```
Comment (fanglong-alliance) [inline csv_permissions/permissions.py]: "own" is still listed here

Comment (fanglong-alliance) [reply csv_permissions/permissions.py]: also the custom line can use an update too

File: csv_permissions/permissions.py [hunk 26]
```diff
@@ -1,174 +1,56 @@
...
+    Dict[PermName, Dict[UserType, UnresolvedEvaluator]],
+    Iterable[str],
+]:
     """
-    Parses the CSV of group permissions returns a list of permissions for further processing.
+    Parses the CSV of user_type permissions returns data for further processing.
 
     CSV file format:
-        entity, app_name, permission_name, is_global, group1, group2, .. groupN
+        entity, app_name, permission_name, is_global, user_type1, user_type2, .. user_typeN
```
Comment (fanglong-alliance) [inline csv_permissions/permissions.py]: header not mentioned

File: csv_permissions/permissions.py [hunk 27]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
...
-            raise RuntimeError(f"Invalid csv_permissions CSV column headers found in {file_path}")
+        prelim_headers = ["Model", "App", "Action", "Is Global"]
+        prelim_header_count = len(prelim_headers)
+
+        if fieldnames[:prelim_header_count] != prelim_headers:
+            raise ValueError(f"Invalid csv_permissions CSV column headers found in {file_path}")
+
+        user_type_headers = fieldnames[prelim_header_count:]
+
+        nonempty_user_type_headers = [user_type for user_type in user_type_headers if user_type != ""]
```
Comment (fanglong-alliance) [inline csv_permissions/permissions.py:85]: why do we want to tolerate the ending comma? assuming this's what its about ... unless you for some reasons want to also accept empty ones in middle of the file?

File: csv_permissions/permissions.py [hunk 28]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
...
+        if len(set(nonempty_user_type_headers)) != len(nonempty_user_type_headers):
+            raise ValueError(f"Duplicate csv_permissions CSV column headers found in {file_path}")
```
Comment (fanglong-alliance) [inline csv_permissions/permissions.py]: probably overkill, but would be more nuanced to show what duplicates are

File: csv_permissions/permissions.py [hunk 29]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
...
-            for auth_group in auth_groups:
-                access_level = row[auth_group]
-
-                perm_group_details[rule_name][auth_group] = PartiallyResolvedPermission(
-                    app_config=app_config,
-                    model=model,
-                    action=action,
-                    access_level=access_level,
-                )
+                raise ValueError("Invalid value for Is Global.")
```
Comment (fanglong-alliance) [inline csv_permissions/permissions.py]: > for Is Global, should be either "yes" or "no"

File: csv_permissions/permissions.py (hunk 12 above)
Comment (fanglong-alliance) [reply csv_permissions/permissions.py:141]: second this question, find it more confusing than helpful (harder for a household cat)

File: README.md [hunk 30]
```diff
@@ -1,4 +1,253 @@
...
+
+            autoreload_started.connect(add_csv_permissions_watcher)
+```
+
+If you're using [`runserver_plus`](https://django-extensions.readthedocs.io/en/latest/runserver_plus.html)
+from `django-extensions` you can add your CSV files to
+[`settings.RUNSERVER_PLUS_EXTRA_FILES`](https://django-extensions.readthedocs.io/en/latest/runserver_plus.html#configuration).   
+
+  
+### The CSV File
```
Comment (fanglong-alliance) [inline README.md:67]: explain how multiple CSV files work, the load order (does not) matter etc

File: csv_permissions/permissions.py [hunk 31]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
...
+        # merge unresolved permissions
+        for permission, new_user_type_to_unresolved in new_permission_to_user_type_to_unresolved.items():
+            if permission not in permission_to_user_type_to_unresolved:
+                permission_to_user_type_to_unresolved[permission] = {}
+            for user_type, new_unresolved in new_user_type_to_unresolved.items():
+                if user_type in permission_to_user_type_to_unresolved[permission]:
+                    if new_unresolved != permission_to_user_type_to_unresolved[permission][user_type]:
+                        raise ValueError(
+                            f"Permission {permission} for user type {user_type} in "
+                            f"{file_path} is inconsistent with a previous CSV file"
```
Comment (fanglong-alliance) [inline csv_permissions/permissions.py]: is it possible for this and prev one to point out which previous csv file? (ie. a temp mapping to record which one gave what)

File: csv_permissions/permissions.py [hunk 32]
```diff
@@ -182,168 +64,253 @@ def _parse_csv(
...
                 )
-            except RuntimeError as e:
-                raise RuntimeError(f"{e} Permission: {permission}")
+                permissions_paths = settings.CSV_PERMISSIONS_PATHS
 
-    return permission_to_group_to_function, permission_is_global
+        # make sure it's always a tuple so that it's hashable and _resolve_functions() can have @lru_cache() applied
+        if not isinstance(permissions_paths, tuple):
+            permissions_paths = tuple(permissions_paths)
+            settings.CSV_PERMISSIONS_PATHS = permissions_paths
```
Comment (fanglong-alliance) [inline csv_permissions/permissions.py:296]: i think maybe either support it to be a not iterable and turn it into a tuple, or throw when its not iterable - mainly to catch people doing settings.CSV_PERMISSIONS_PATHS="x.csv"

File: csv_permissions/permissions.py (hunk 13 above)
Comment (fanglong-alliance) [reply csv_permissions/permissions.py]: ^ ya, means if it's vanilla group we can still do

File: doc/background.md [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+      permissions system by allowing for permissions to be based on execution of code
+      
+* **No versioning**. By default, making changes to permissions leaves no audit trail.
+
+* **Permission names**. Django permission names are in the form `f'{app_label}.{action}_{model}'`.
+    This does not sort well.
+
+Default django permissions has the following tradeoffs: 
+
+* **Enumerating permssions**:
```
Comment (fanglong-alliance) [inline doc/background.md]: typo

File: csv_permissions/permissions.py (hunk 2 above)
Comment (levic) [reply csv_permissions/permissions.py]: updated (also the other params for this function)

File: README.md (hunk 3 above)
Comment (levic) [reply README.md]: added to README

File: README.md (hunk 5 above)
Comment (levic) [reply README.md:104]: Added link

File: README.md (hunk 6 above)
Comment (levic) [reply README.md]: Depends on what you want to do -- it's left entirely up to the application.

`csv_permissions.legacy.resolve_custom_evaluator` can be used to get the old behaviour, but the intention is that other than the core yes/all/'' evaluators you'll define them in your application.

I anticipate that at some point we'll have a built-in `rule:foo` that will check that `foo` is a django-rules predicate but didn't want to do this until I'd used it on a project to make sure that the interface is settled (eg, what do we do about the same rule name in 2 different apps? we can fully qualify it but that's going to make the cells contents quite long -- do we want to allow unqualified rules if they're unambiguous? perhaps is better to provide a django-rules evaluator in csv_permissions but not make it default, and let the app to decide what to do)

File: README.md (hunk 7 above)
Comment (levic) [reply README.md:189]: Have clarified that this is the default setting if you don't set it

File: README.md (hunk 8 above)
Comment (levic) [reply README.md]: > could you have a case where you want a custom non-model permission that does accept an object of some kind?

I hadn't thought of that; conceivably yes. Have added a comment that this restriction may be relaxed in future.

File: csv_permissions/evaluators.py (hunk 10 above)
Comment (levic) [reply csv_permissions/evaluators.py:54]: That used to be allowed, but it meant that there was two different ways to express the same thing.

I found it easier to scan big files if empty cells meant "no permissions".

File: csv_permissions/evaluators.py (hunk 9 above)
Comment (levic) [reply csv_permissions/evaluators.py:22]: (see above; added comment, may relax restriction in future)

File: csv_permissions/legacy.py (hunk 11 above)
Comment (levic) [reply csv_permissions/legacy.py]: removed

File: csv_permissions/permissions.py (hunk 12 above)
Comment (levic) [reply csv_permissions/permissions.py:141]: If you have trailing commas in your CSV file then the last column will be read as an empty header.

It was only slightly more work to then allow empty columns in the middle of the file too, but means that you can break up your users if you want (eg in the current project I'm working on we have 3 user profiles, each with 2-3 user_types in each profile so leaving a blank column can make it a little easier to visually break them up)

File: csv_permissions/permissions.py (hunk 13 above)
Comment (levic) [reply csv_permissions/permissions.py]: There is now a `CSV_PERMISSIONS_GET_USER_TYPE` setting that defaults to returning a function that resolves to `return user.user_type`, so this has been removed

File: README.md (hunk 14 above)
Comment (levic) [reply README.md:99]: added a comment inside the CSV file

File: README.md (hunk 15 above)
Comment (levic) [reply README.md]: Added a link to User Type section and the following extra para:

> User Types are the `csv_permissions` equivalent of django's user Group. A user is
> assigned a user type and a user type is assigned a set of permissions in the CSV file.

File: README.md (hunk 16 above)
Comment (levic) [reply README.md]: fixed

File: README.md (hunk 21 above)
Comment (levic) [reply README.md]: fixed

File: csv_permissions/permissions.py (hunk 32 above)
Comment (levic) [reply csv_permissions/permissions.py:296]: (Leaving this as a placeholder comment, please don't resolve this comment yet)

Every single time you do a permissions check django instantiates a new copy of the permissions backend. When the backend does nothing this is fine, but there's a lot csv permissions needs to do to load the files. I've worked around this by putting an `@lru_cache` on `_resolve_functions` and having the constructor for `CsvPermissionsBackend` do as little as possible. If we allow both settings then we either need to do extra work on each CsvPermissionsBackend init, or we need to have some way to ensure that a cached CSV_PERMISSIONS_PATHS doesn't override any updates to CSV_PERMISSIONS_PATH in the tests.

File: csv_permissions/permissions.py (hunk 31 above)
Comment (levic) [reply csv_permissions/permissions.py]: Note to self: could add a `source` attribute to `UnresolvedEvaluator`

File: README.md (hunk 30 above)
Comment (levic) [reply README.md:67]: added comment

File: csv_permissions/permissions.py (hunk 29 above)
Comment (levic) [reply csv_permissions/permissions.py]: updated

File: csv_permissions/permissions.py (hunk 27 above)
Comment (levic) [reply csv_permissions/permissions.py:85]: depending on the CSV editor you're using you sometimes don't get a choice, it will insert a closing comma whether you want it to or not

File: csv_permissions/permissions.py (hunk 28 above)
Comment (levic) [reply csv_permissions/permissions.py]: added

edit: added _first_ duplicate header

File: README.md (hunk 6 above)
Comment (levic) [reply README.md]: have removed it to keep things simple; it shouldn't have been in there

File: README.md (hunk 20 above)
Comment (levic) [reply README.md]: fixed

File: README.md (hunk 1 above)
Comment (levic) [reply README.md]: (removed)

File: csv_permissions/permissions.py (hunk 27 above)
Comment (fanglong-alliance) [reply csv_permissions/permissions.py:85]: use a better editor! 🤣

File: csv_permissions/evaluators.py (hunk 10 above)
Comment (fanglong-alliance) [reply csv_permissions/evaluators.py:54]: ahahahaha, and its not just me and dave want a "no" - ash the newcomer just left this msg in his training:

```
For the admin/customer permissions (the last couple of columns in the CSV), is there a reason we can't specify "no" and we're instead forced to just leave that cell blank?

I had "no" in them and couldn't figure out why it was throwing me errors, I ended up reading through the permission processing file to figure it out. (https://github.com/AllianceSoftware/django-csvpermissions/blob/master/csv_permissions/permissions.py#L91)
 

I know that in the training MD file, it does say "Blank cells are equivalent to that user type having no permission", but I interpreted that as "a cell could either be blank OR no" and I figured at the time that reading no's in the CSV would be easier, rather than intermittent blank cells.
```

so, really, "no" is what people _would_ be expecting 😆, in addition to being much more friendlier to `cat`

Comment (levic) [reply csv_permissions/evaluators.py:54]: okay, will change this.

@fanglong-alliance @davecoates this raises the question of what a blank cell means:

- fall back to "no" permissions
- "unspecified"
 
where this matters if you have multiple csv files and one is blank and the other is "yes", right now blank != yes and it throw an error but if we treat it as "unspecified" then the "yes" should win out and not throw an error

my inclination is the latter

File: README.md (hunk 18 above)
Comment (levic) [reply README.md:127]: see above: will change to "no"

File: csv_permissions/permissions.py (hunk 25 above)
Comment (levic) [reply csv_permissions/permissions.py]: have made DRY: removed docs here and referenced README.md

File: csv_permissions/permissions.py (hunk 26 above)
Comment (levic) [reply csv_permissions/permissions.py]: see above: references the README now

File: README.md (hunk 17 above)
Comment (levic) [reply README.md:121]: Added an extra para clarifying a little below this:

```
The distinction between `all` and `yes` is that `all` is a per-object
permission and `yes` is a [global permission](#global-permissions). 
```

File: README.md (hunk 23 above)
Comment (levic) [reply README.md:316]: Added an extra para:

```
Note the handling of the case where a permission has no model.
Examples of this can be seen in the `report_outstanding` and `report_popularity`
permissions in the [sample CSV file](#the-csv-file).
```

File: csv_permissions/permissions.py (hunk 31 above)
Comment (levic) [reply csv_permissions/permissions.py]: It now reports a previous file that is inconsistent with the current one (also for the is_global check a little earlier)

File: csv_permissions/permissions.py (hunk 32 above)
Comment (levic) [reply csv_permissions/permissions.py:296]: Have changed it so that `CSV_PERMISSION_PATHS` (list/tuple of paths) and `CSV_PERMISSION_PATH` (single path) are now both allowed

File: csv_permissions/evaluators.py (hunk 10 above)
Comment (fanglong-alliance) [reply csv_permissions/evaluators.py:54]: yes, i think that would be how i'd expect it to work. this way it also opens up for people to, if they really want, define a single row of permissions in a file w/o worrying it conflict w/ others

**Comment by fanglong-alliance** [comment]
awesome, think that makes all my comments resolved

File: README.md (hunk 1 above)
Comment (levic) [reply README.md]: Have removed from the instructions as it doesn't do anything right now, but the app config class still exists (even though it doesn't do anything except set the app name)

File: csv_permissions/evaluators.py (hunk 10 above)
Comment (levic) [reply csv_permissions/evaluators.py:54]: implemented

**Comment by davecoates** [comment]
Looks good 👍

**Comment by levic** [comment]
This resolves #2  #3 #5
